signer/src/block_notifier.rs

```diff
@@ -0,0 +1,33 @@
+//! This module defines traits for observing and notifying about Bitcoin block headers.
+//!
+//! The `BlockObserver` trait should be implemented by any component that wants to receive
+//! notifications of new block headers. The `BlockNotifier` trait should be implemented by any
+//! component that can notify subscribed observers about new block headers.
+
+use std::future::Future;
+use std::sync::Arc;
+use std::error::Error;
+
+/// The `BlockObserver` trait defines a method to be notified of new block headers.
+pub trait BlockObserver: Send + Sync {
+    /// Errors occurring during notification
+    type Error: Error;
+
+    /// Notify the observer of a new block header
+    /// This method is called by the `BlockNotifier` whenever a new block header is detected
+    fn notify(&self, block_hash: &bitcoin::BlockHash) -> Result<(), Self::Error>;
+}
+
+/// The `BlockNotifier` trait defines methods for subscribing observers and running the notifier.
+pub trait BlockNotifier {
+    /// Errors occurring during subscription or running the notifier
+    type Error: Error;
+
+    /// Adds the provided observer to the list of observers that will be notified of new block headers
+    fn subscribe(&mut self, observer: Arc<dyn BlockObserver<Error = Self::Error>>) -> impl Future<Output = Result<(), Self::Error>> + Send;
```

Comment:
I think this could be simplified so that the `BlockNotifier` won't have to be aware of the `BlockObserver`. This could just return a `tokio::mpsc::Receiver` or a [`futures_core::stream::Stream`](https://docs.rs/futures-core/0.3.30/futures_core/stream/trait.Stream.html) of blocks (though the latter would probably require an associated type for the actual stream that implements the `Stream` trait).

Reply:
@netrome  I got rid of the BlockObserver trait and since the BlockNotifier trait was small enough, I added the impl with Electrum in same PR. Is this closer to what we want?

I made the integ test to work locally by changing Gowtham's electrs docker details to use 0.0.0.0 but unfortunately, I couldn't make it work with the original 121.0.0.1 host so I didn't push the changed devenv files and ignored the test for now.

Reply:
Okay, I think it should be using "0.0.0.0" within the container so anything else is probably a bug. Could you include the fix in the PR?